### PR TITLE
cassandra 3.x - support for SMALLINT, TINYINT, DATE

### PR DESCRIPTION
**Pull request by pawelrychlik**
With the 3.0 release of Cassandra, there have been added some new data types: `tinyint` and `smallint` (https://docs.datastax.com/en/cql/3.1/cql/cql_reference/cql_data_types_c.html).

If you tried to use prestodb to work with a Cassandra 3.x table that has columns of that new types, you would run into a `NullPointerException: cassandraType is null`.

This PR adds support for these two types.

This little test scenario might come in handy:
```
cqlsh> create table test.presto_types (id text, small smallint, tiny tinyint, PRIMARY KEY(id));

presto:test> select * from test.presto_types;

Query 20170117_135445_00006_68jce failed: cassandraType is null
java.lang.NullPointerException: cassandraType is null
	at java.util.Objects.requireNonNull(Objects.java:228)
	at com.facebook.presto.cassandra.CassandraColumnHandle.<init>(CassandraColumnHandle.java:62)
```

### CODE REVIEW

**Comment by ArturGajowy** [comment]
Please rephrase commit messages according to http://chris.beams.io/posts/git-commit/. Mainly,
- start with a capital letter
- use imperative style ('add' isntead of 'adding')

Also, please capitalize the type names and mention that the change is scoped to cassandra connector. ("e.g. Add support for TINYINT and SMALLINT types in Cassandra connector" - probably should be shorter to fit in 70 chars)

**Comment by ArturGajowy** [comment]
The code looks good, btw.

**Comment by pawelrychlik** [comment]
Well, I realised that, it's not really a low-hanging fruit, is it? :)

Now I should probably bump up cassandra-version to 3.x in the connector's `pom.xml` (what, latest 3.9?), which in turn requires `cassandra.yaml` changes (and prolly much more than that), and might have large impact overall.

@cawallin how do we go about that? anyone up for bumping it up? :)

**Comment by cawallin** [comment]
@pawelrychlik -- we're currently working on a PR to remove the Thrift client from the presto-cassandra connector (https://github.com/prestodb/presto/pull/7013). Until that goes in, the latest version of Cassandra won't work with Presto since the Thrift client is no longer supported in Cassandra. So I guess this PR will have to wait, I didn't catch that this is only supported in the latest version of Cassandra :) You should still work on tests, etc; we're hoping to have the Thrift change in soon.

**Comment by pawelrychlik** [comment]
I've actually added DATEs here as well. My next step would be to extend the SmokeTest to cover that type too. The problem is that the tests won't work with Cassandra 2.1.6 now.

There's not so much documentation around, so I had to sort things out by myself with a lot of trial & error involved. Therefore I'd appreciate a sanity check / code review, especially for the DATE parts. Thanks!.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordCursor.java [hunk 1]
```diff
@@ -111,6 +111,8 @@ public long getLong(int i)
                 return currentRow.getLong(i);
             case TIMESTAMP:
                 return currentRow.getTimestamp(i).getTime();
+            case DATE:
```
Comment (cawallin) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordCursor.java]: Like for the other types, you probably need to update CassandraRecordSink for CTAS.

**Comment by cawallin** [comment]
@pawelrychlik We got https://github.com/prestodb/presto/pull/7013 merged, so Cassandra 3.x is fully supported. Go ahead and rebase on master and add some tests!

**Comment by pawelrychlik** [comment]
@cawallin according to https://github.com/prestodb/presto-cassandra-server/blob/master/pom.xml#L46 cassandra is still at 2.1.16, isn't it?

**Comment by cawallin** [comment]
@pawelrychlik We want to support both 2.1.x and 3.x, and so currently we run the Cassandra integration tests against in-memory 2.1.16 Cassandra and the product tests against 3.x Cassandra in Docker. You can write product testse for smallint/tinyint/date. Let me know if you need more pointers for that.

**Comment by eparesh** [comment]
I am using cassandra 3.11.2, is this issue fixed?

File: presto-cassandra/src/test/java/com/facebook/presto/cassandra/TestCassandraType.java [hunk 2]
```diff
@@ -32,6 +32,8 @@ public void testJsonMapEncoding()
         assertTrue(isValidJson(CassandraType.buildArrayValue(Lists.newArrayList(1, 2, 3), CassandraType.INT)));
         assertTrue(isValidJson(CassandraType.buildArrayValue(Lists.newArrayList(100000L, 200000000L, 3000000000L), CassandraType.BIGINT)));
         assertTrue(isValidJson(CassandraType.buildArrayValue(Lists.newArrayList(1.0, 2.0, 3.0), CassandraType.DOUBLE)));
+        assertTrue(isValidJson(CassandraType.buildArrayValue(Lists.newArrayList((short) -32768, (short) 0, (short) 32767), CassandraType.SMALLINT)));
+        assertTrue(isValidJson(CassandraType.buildArrayValue(Lists.newArrayList((byte) -128, (byte) 0, (byte) 127), CassandraType.TINYINT)));
```
Comment (kokosing) [inline presto-cassandra/src/test/java/com/facebook/presto/cassandra/TestCassandraType.java]: Can you please also update `com.facebook.presto.tests.cassandra.Select#testAllDatatypes`?

**Comment by dowjones226** [comment]
Hi, I am also using cassandra 3.11.2, is this issue fixed yet?

**Comment by ebyhr** [comment]
@pawelrychlik Gentle reminder :) I can take over this if you are busy. Some people want to use these types.

**Comment by pawelrychlik** [comment]
@ebyhr I'd be happy if you could take over. I don't want to block anyone, but I'm no longer working with prestodb, so I'd have to set it up and dig into it again.

**Comment by ebyhr** [comment]
@kokosing Current cassandra-server's version for embedded server is [2.1.16-1](https://github.com/prestodb/presto/blob/master/pom.xml#L1062), but smallint and tiny are introduced from [2.2.0-rc1](https://datastax-oss.atlassian.net/browse/JAVA-786). Travis was failing with this compatibility since those types doesn't exist. Therefore, we need to upgrade [presto-cassandra-server](https://github.com/prestodb/presto-cassandra-server) or add only product test. 

Additionally, it may be necessary to upgrade cassandra [protocol version](https://github.com/prestodb/presto/blob/173216dc3517e8531beb5f7d218fd543931b4952/presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClientModule.java#L83) from v3 to v4. 
There's already PR (https://github.com/prestodb/presto/issues/7943) which users specify protocol version.

**Comment by mnmami** [comment]
Where are you on this? I need to access Cassandra table with date type column.

**Comment by ebyhr** [comment]
@mnmami DATE type was also introduced in the Cassandra 2.2 release. I can send new PR which added product test to pawelrychlik's commit after upgrading  presto-cassandra-server and #7943 is merged.

**Comment by mnmami** [comment]
@ebyhr thanks, sounds good -- any estimated time line?

**Comment by mnmami** [comment]
Six months later, and this still hasn't been picked up. Sad!

**Comment by kokosing** [comment]
@ebyhr or @pawelrychlik Can you please rebase?

**Comment by ebyhr** [comment]
@kokosing I will rebase later.

**Comment by ebyhr** [comment]
I could test additional column types with cassandra 2.2.0, but it requires to update protocol-version from V3(default) to V4. I think there are two options. 
1. presto-cassandra test keeps using V3, presto-product-test uses V4
→ No test for additional columns on cassandra 2.x, test only with 3.x)

2. both presto-cassandra test and presto-product-test use V4 
→ Don't test with V3 anymore

What do you think about it? @kokosing

**Comment by findepi** [comment]
> I think we could move forward with V4. Cassandra 2.2.0 (https://github.com/apache/cassandra/releases/tag/cassandra-2.2.0) was released over 3 years ago, so it seems safe to upgrade.

It depends on the community adoption velocity and other factors.
For example, Hive 1.2 was released I don't remember when, but It wouldn't be safe to drop support for it.

**Comment by kubum** [comment]
Hey @findepi! What is the best way of handling it? Would it be possible to handle date and other data types depending on a discovered version?

**Comment by ebyhr** [comment]
This is a branch I am rebasing.
https://github.com/ebyhr/presto/commit/fef6ba2382bf7f8d2395d7b9e27ac18d1257d8c9

I think we need to additional logic to support V3 protocol like below. Without this logic, CTAS from tpch on smoke integration test fails since the table has DATE type. Currently, the DATE column is defined as TEXT.
https://github.com/ebyhr/presto/commit/fef6ba2382bf7f8d2395d7b9e27ac18d1257d8c9#diff-c68d6013815cb7c117bcb89d7d70537dR559
https://github.com/ebyhr/presto/commit/fef6ba2382bf7f8d2395d7b9e27ac18d1257d8c9#diff-cb3c31d16b6364fa38b00109e14261a9R145

**Comment by kokosing** [comment]
@ebyhr Sounds good.

Do we have now product tests running on V3? If so we could add V4 to product tests so we would run them in two configurations V3 and V4. Also we need to make sure that we test these new types there. Then presto-cassandra tests could use V4 (be closer to the edge), while the product tests with V3 would test the legacy path.

**Comment by ebyhr** [comment]
@kokosing  Yes, product test is running on V3. Actually, I had prepared presto-cassandra with V3 and product test with V4.

There are additional tasks to run V4 on presto-cassandra.
1. Release new presto-cassandra-server to maven repo https://github.com/prestodb/presto-cassandra-server/pull/4
2. Look for the way to refresh size_estimates. Whereas current logic calls it via MBean, 2.2.0 doesn't have it. I haven't found alternate method yet... https://github.com/prestodb/presto/blob/3060c65a1812c6c8b0c2ab725b0184dbad67f0ed/presto-cassandra/src/test/java/com/facebook/presto/cassandra/EmbeddedCassandra.java#L188-L198

**Comment by ppapishe** [comment]
Hello guys, Just want to know if there is an ETA on this.

**Comment by ebyhr** [comment]
@ppapishe Hi, I'm migrating this PR to https://github.com/prestosql/presto/pull/141. ETA is unsure, but I will focus this to ship within 1 month.

**Comment by ppapishe** [comment]
> @ppapishe Hi, I'm migrating this PR to [prestosql/presto#141](https://github.com/prestosql/presto/pull/141). ETA is unsure, but I will focus this to ship within 1 month.

Thank you very much @ebyhr . This is an amazing tool and Thank you all for all the good work.

**Comment by pawelrychlik** [comment]
Continued in https://github.com/prestosql/presto/pull/141
Closing this one.

**Comment by ebyhr** [comment]
@eparesh @dowjones226  @mnmami @kubum @ppapishe Sorry for waiting you. The PR was merged into https://github.com/prestosql/presto. It will be released on 303.

**Comment by ppapishe** [comment]
> @eparesh @dowjones226 @mnmami @kubum @ppapishe Sorry for waiting you. The PR was merged into https://github.com/prestosql/presto. It will be released on 304.

Thank you very much

**Pull request closed without merging**